lib/streamlit/elements/lib/dialog.py

```diff
@@ -0,0 +1,108 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import time
+from types import TracebackType
+from typing import Literal, cast
+
+from typing_extensions import TypeAlias
+
+from streamlit.cursor import Cursor
+from streamlit.delta_generator import DeltaGenerator, _enqueue_message
+from streamlit.proto.Block_pb2 import Block as BlockProto
+from streamlit.proto.ForwardMsg_pb2 import ForwardMsg
+
+DialogWidth: TypeAlias = Literal["small", "large"]
+
+
+class Dialog(DeltaGenerator):
+    @staticmethod
+    def _create(
+        parent: DeltaGenerator,
+        title: str,
+        *,
+        dismissible: bool = True,
+        width: DialogWidth = "small",
+    ) -> Dialog:
+        block_proto = BlockProto()
+        block_proto.dialog.title = title
+        block_proto.dialog.dismissible = dismissible
+        block_proto.dialog.width = str(width)
+
+        delta_path: list[int] = (
+            parent._active_dg._cursor.delta_path if parent._active_dg._cursor else []
+        )
```

Comment:
I'm perhaps a bit nervous that we are inheriting some delta generator logic here (in multiple places). I won't block on this, but do we have a good story on how to organize the delta path/delta generator so that it's not likely to cause issues and errors on refactors/changes.

Reply:
Yeah great point! I feel the same way and think that in general we decouple this better. Also, having the cyclic dependencies between the DG and components feels a little bit weird. Maybe we can brainstorm about improvements for this once I am back, and maybe @LukasMasuch already has thoughts about this as this follows a similar approach as `st.expander`.

Reply:
Just bookmarking for a conversation to be had here. Bare minimum, we should have a comment explaining this logic.

Reply:
I think I'm getting this now. Do we have to worry where fragments edit regions outside and can therefore change the delta_path?

Reply:
Perhaps, as I believe some feedback for fragments has already shown that editing outside-regions can be problematic. I guess its related to the overall out-of-tree delta generators etc. We definitely should keep it on our radar, especially because expander etc. would also be impacted by an issue here. I suggest to file a tech debt as well as this goes beyond just dialog and keep an eye out for potential issues in the meantime.